components/tidb_query/src/codec/mysql/decimal.rs

```
@@ -1070,25 +1070,35 @@ impl Decimal {
 
     /// convert_to(ProduceDecWithSpecifiedTp in tidb)
     /// produces a new decimal according to `flen` and `decimal`.
+    /// # panic
+    /// - flen should >=decimal, so panic when flen < decimal
     pub fn convert_to(self, ctx: &mut EvalContext, flen: u8, decimal: u8) -> Result<Decimal> {
+        assert!(flen >= decimal);
         let (prec, frac) = self.prec_and_frac();
         if !self.is_zero() && prec - frac > flen - decimal {
-            return Ok(max_or_min_dec(self.negative, flen, decimal));
-            // TODO:select (cast 111 as decimal(1)) causes a warning in MySQL.
-        }
-
-        if frac == decimal {
-            return Ok(self);
-        }
-
-        let tmp = self.clone();
-        let ret = self.round(decimal as i8, RoundMode::HalfEven).unwrap();
-        // TODO: process over_flow
-        if !ret.is_zero() && frac > decimal && ret != tmp {
-            // TODO handle InInsertStmt in ctx
-            ctx.handle_truncate(true)?;
+            // select (cast 111 as decimal(1)) causes a warning in MySQL.
+            ctx.handle_overflow(Error::overflow(
+                "Decimal",
+                &format!("({}, {})", flen, decimal),
+            ))?;
+            Ok(max_or_min_dec(self.negative, flen, decimal))
+        } else if frac != decimal {
+            let old = self.clone();
+            // TODO, is unwrap ok?
+            let rounded = self.round(decimal as i8, RoundMode::HalfEven).unwrap();
```

Maybe we should propagate the error to the caller.